homeassistant/components/solax/sensor.py

```diff
@@ -14,37 +14,39 @@
     SensorEntity,
     SensorStateClass,
 )
-from homeassistant.const import CONF_IP_ADDRESS, CONF_PORT, TEMP_CELSIUS
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD, CONF_PORT, TEMP_CELSIUS
 from homeassistant.core import HomeAssistant
 from homeassistant.exceptions import PlatformNotReady
 import homeassistant.helpers.config_validation as cv
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.event import async_track_time_interval
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
-DEFAULT_PORT = 80
+from .const import DOMAIN
 
+DEFAULT_PORT = 80
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {
         vol.Required(CONF_IP_ADDRESS): cv.string,
         vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    }
+    },
 )
-
 SCAN_INTERVAL = timedelta(seconds=30)
 
 
-async def async_setup_platform(
+async def async_setup_entry(
     hass: HomeAssistant,
-    config: ConfigType,
+    entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
-    """Platform setup."""
-    api = await real_time_api(config[CONF_IP_ADDRESS], config[CONF_PORT])
-    endpoint = RealTimeDataEndpoint(hass, api)
+    """Entry setup."""
+    api = await real_time_api(
+        entry.data[CONF_IP_ADDRESS], entry.data[CONF_PORT], entry.data[CONF_PASSWORD]
+    )
```

Comment:
I  would recommend moving this up into the integration config entry setup.
That would allow for handling errors (e.g., could not connect right now); if we throw a ConfigEntryNotReady error at that point, Home Assistant will automatically try to reconnect/re-setup.

Reply:
I already do a connection validation on config entry setup with the function validate_api.

Reply:
You don't?

![image](https://user-images.githubusercontent.com/195327/150778553-ed55351d-467b-42c8-8eab-87b053b52dc1.png)

Reply:
Sorry, I thought you were talking about **async_step_user** in **config_flow.py**

Reply:
Is this what you had in mind?

`    try:
        await validate_api(entry.data)
    except Exception as err:
        raise ConfigEntryNotReady from err

    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
    return True`

Reply:
No, `validate_api` is something from the config flow. That does not apply here.

Instead, try set up the connection here, check if it works. If not raise a ConfigEntryNotReady.

Store the API client in hass data and pick it up again in the sensor for further use.